quickwit/quickwit-search/src/leaf.rs

```diff
@@ -394,15 +398,195 @@ async fn leaf_search_single_split(
 ///
 /// This include things such as sorting result by a field or _score when no document is requested,
 /// or applying date range when the range covers the entire split.
-fn rewrite_request(search_request: &mut SearchRequest, split: &SplitIdAndFooterOffsets) {
+fn rewrite_request(
+    search_request: &mut SearchRequest,
+    split: &SplitIdAndFooterOffsets,
+    timestamp_field: Option<&str>,
+) {
     if search_request.max_hits == 0 {
         search_request.sort_fields = Vec::new();
     }
+    if let Some(timestamp_field) = timestamp_field {
+        remove_redundant_timestamp_range(search_request, split, timestamp_field);
+    }
     rewrite_start_end_time_bounds(
         &mut search_request.start_timestamp,
         &mut search_request.end_timestamp,
         split,
-    )
+    );
+}
+
+/// remove timestamp range that would be present both in QueryAst and SearchRequest
+///
+/// this can save us from doing double the work in some cases, and help with the partial request
+/// cache.
+fn remove_redundant_timestamp_range(
+    search_request: &mut SearchRequest,
+    split: &SplitIdAndFooterOffsets,
+    timestamp_field: &str,
+) {
+    let Ok(query_ast) = serde_json::from_str(search_request.query_ast.as_str()) else {
+        // an error will get raised a bit after anyway
+        return;
+    };
+
+    let start_timestamp = match (search_request.start_timestamp, split.timestamp_start) {
```

Comment:
so one thing that is sad with this approach, is that we end up with different execution logic if 
the timestamp is passed in the query , or in the search query parameter.

Could we remove all of the "must" (what you did on the bool query is good) timestamp range queries, and identify the start end timestamp range it suggests, intersect the start timestamp given by the request parameters, and then  build the fastfield field off of this values?

Reply:
I'm not totally sure I understand what you mean. In-body ranges have a higher precision than query parameters so we can't just remove them.
Or do you mean removing all the must `RangeQuery`, stop using `{start,end}_timestamp` in the collector, and instead add a single `RangeQuery` at the root of the ast?